File: src/session_lock.c
/* SPDX-License-Identifier: MIT
 *
 * Screen locking.
 *
 * Without ext-session-lock-v1 there is no way to lock the session at all:
 * swaylock and every other locker binds this protocol and exits immediately if
 * it is missing, so the screen simply never locks.
 *
 * The protocol exists because a lock screen has a security property no
 * ordinary client can be trusted with: once the compositor acknowledges a
 * lock, nothing else may be shown or receive input until the locker says
 * otherwise — and if the locker *crashes*, the session must stay locked rather
 * than falling open. That is why the lock surfaces live on their own layer
 * above everything, and why a crashed locker leaves a blank screen instead of
 * revealing the desktop.
 */
#define _POSIX_C_SOURCE 200809L

#include <stdlib.h>

#include <wlr/types/wlr_output_layout.h>
#include <wlr/types/wlr_scene.h>
#include <wlr/types/wlr_session_lock_v1.h>
#include <wlr/util/log.h>

#include "viewport-view.h"
#include "viewport-input.h"
#include "viewport-output.h"

struct viewport_lock_surface {
	struct wlr_session_lock_surface_v1 *lock_surface;
	struct viewport_output *output;
	struct wlr_scene_tree *tree;

	struct wl_listener map;
	struct wl_listener destroy;
};

static void handle_lock_surface_map(struct wl_listener *listener, void *data)
{
	struct viewport_lock_surface *surface =
		wl_container_of(listener, surface, map);
	struct viewport_server *server = surface->output->server;

	/* Focus the locker so typing a password reaches it rather than whatever
	 * was focused when the screen locked. */
	struct wlr_keyboard *keyboard = wlr_seat_get_keyboard(server->seat);
	wlr_seat_keyboard_notify_enter(server->seat,
		surface->lock_surface->surface,
		keyboard != NULL ? keyboard->keycodes : NULL,
		keyboard != NULL ? keyboard->num_keycodes : 0,
		keyboard != NULL ? &keyboard->modifiers : NULL);
}

static void handle_lock_surface_destroy(struct wl_listener *listener,
	void *data)
{
	struct viewport_lock_surface *surface =
		wl_container_of(listener, surface, destroy);

	wl_list_remove(&surface->map.link);
	wl_list_remove(&surface->destroy.link);
	free(surface);
}

static void handle_new_lock_surface(struct wl_listener *listener, void *data)
{
	struct viewport_lock *lock = wl_container_of(listener, lock, new_surface);
	struct wlr_session_lock_surface_v1 *lock_surface = data;
	struct viewport_output *output = lock_surface->output->data;
	if (output == NULL) {
		/* The monitor is already going away: handle_output_destroy() clears
		 * wlr_output->data before freeing the struct, so this downcast legally
		 * yields NULL, and every line below dereferences it. wlroots listens on
		 * the same output's destroy signal and will take this lock surface with
		 * it, so dropping it here loses nothing. */
		wlr_log(WLR_DEBUG, "lock surface for a dying output, ignoring");
		return;
	}

	struct viewport_lock_surface *surface = calloc(1, sizeof(*surface));
	if (surface == NULL) {
		return;
	}
	surface->lock_surface = lock_surface;
	surface->output = output;
	surface->tree = wlr_scene_subsurface_tree_create(lock->server->layer_lock,
		lock_surface->surface);

	if (surface->tree != NULL) {
		struct wlr_box box;
		wlr_output_layout_get_box(lock->server->output_layout,
			output->wlr_output, &box);
		wlr_scene_node_set_position(&surface->tree->node, box.x, box.y);
		/* The locker covers exactly one output, so it is told that size. */
		wlr_session_lock_surface_v1_configure(lock_surface, box.width,
			box.height);
	}

	surface->map.notify = handle_lock_surface_map;
	wl_signal_add(&lock_surface->surface->events.map, &surface->map);
	surface->destroy.notify = handle_lock_surface_destroy;
	wl_signal_add(&lock_surface->events.destroy, &surface->destroy);
}

/* A black rectangle under every lock surface, covering the whole layout.
 *
 * The locker draws one surface per output, but it only learns about an output
 * once it exists — so a monitor plugged in while the session is locked would
 * show the desktop until the locker caught up. Covering the entire layout means
 * there is no such window, and it costs one rectangle.
 *
 * It also covers the case of a locker that simply never draws on some output,
 * which the protocol allows and which would otherwise leave that screen
 * readable. */
static void update_backdrop(struct viewport_server *server)
{
	/* Keyed on the session being locked rather than on a lock object existing:
	 * after a locker dies there is no lock object and the screen still has to
	 * stay covered. */
	if (!server->locked && server->lock == NULL) {
		return;
	}

	struct wlr_box layout;
	wlr_output_layout_get_box(server->output_layout, NULL, &layout);

	if (server->lock_backdrop == NULL) {
		const float black[4] = { 0.0f, 0.0f, 0.0f, 1.0f };
		server->lock_backdrop = wlr_scene_rect_create(server->layer_lock,
			layout.width, layout.height, black);
		if (server->lock_backdrop == NULL) {
			return;
		}
		/* Below the lock surfaces, which are added to the same tree. */
		wlr_scene_node_lower_to_bottom(&server->lock_backdrop->node);
	} else {
		wlr_scene_rect_set_size(server->lock_backdrop, layout.width,
			layout.height);
	}

	wlr_scene_node_set_position(&server->lock_backdrop->node, layout.x,
		layout.y);
}

/* Called when outputs change, so a screen that appears while locked is covered
 * rather than revealing what is behind. */
void viewport_session_lock_outputs_changed(struct viewport_server *server)
{
	update_backdrop(server);
}

static void handle_unlock(struct wl_listener *listener, void *data)
{
	struct viewport_lock *lock = wl_container_of(listener, lock, unlock);
	struct viewport_server *server = lock->server;

	server->locked = false;
	wlr_scene_node_set_enabled(&server->layer_lock->node, false);

	/* Unlocking is the one path that ends the cover. */
	if (server->lock_backdrop != NULL) {
		wlr_scene_node_destroy(&server->lock_backdrop->node);
		server->lock_backdrop = NULL;
	}

	/* Hand the keyboard back to whatever was focused before. */
	if (server->focused != NULL) {
		struct viewport_toplevel *toplevel = server->focused;
		server->focused = NULL;
		viewport_toplevel_focus(toplevel);
	} else {
		viewport_focus_web(server);
	}

	wlr_log(WLR_INFO, "session unlocked");
}

static void handle_lock_destroy(struct wl_listener *listener, void *data)
{
	struct viewport_lock *lock = wl_container_of(listener, lock, destroy);
	struct viewport_server *server = lock->server;

	wl_list_remove(&lock->new_surface.link);
	wl_list_remove(&lock->unlock.link);
	wl_list_remove(&lock->destroy.link);

	if (server->locked) {
		/* The locker died without unlocking. Staying locked is the whole point
		 * of the protocol: a crash must not expose the desktop.
		 *
		 * The lock layer does not do that by itself. It is a scene tree, and an
		 * empty one draws nothing and occludes nothing — the locker's surfaces
		 * went with the client, so leaving the layer enabled and empty showed
		 * the desktop straight through it, with every window readable and the
		 * shell taking input, while `locked` stayed true and no client could be
		 * clicked again. The backdrop is the only thing actually covering the
		 * screen, so it is what must survive, and update_backdrop() keeps
		 * resizing it for outputs that arrive later.
		 *
		 * It is left deliberately blank. Drawing anything here would be a
		 * password prompt the compositor cannot honour. */
		update_backdrop(server);
		wlr_log(WLR_ERROR,
			"lock client vanished while locked; session stays locked "
			"(screen stays covered; switch VT to recover)");
	} else if (server->lock_backdrop != NULL) {
		wlr_scene_node_destroy(&server->lock_backdrop->node);
		server->lock_backdrop = NULL;
	}

	free(lock);
	server->lock = NULL;
}

static void handle_new_lock(struct wl_listener *listener, void *data)
{
	struct viewport_server *server =
		wl_container_of(listener, server, new_session_lock);
	struct wlr_session_lock_v1 *wlr_lock = data;

	if (server->lock != NULL) {
		/* Already locked by someone else. */
		wlr_session_lock_v1_destroy(wlr_lock);
		return;
	}

	struct viewport_lock *lock = calloc(1, sizeof(*lock));
	if (lock == NULL) {
		wlr_session_lock_v1_destroy(wlr_lock);
		return;
	}
	lock->server = server;
	lock->lock = wlr_lock;
	server->lock = lock;

	lock->new_surface.notify = handle_new_lock_surface;
	wl_signal_add(&wlr_lock->events.new_surface, &lock->new_surface);
	lock->unlock.notify = handle_unlock;
	wl_signal_add(&wlr_lock->events.unlock, &lock->unlock);
	lock->destroy.notify = handle_lock_destroy;
	wl_signal_add(&wlr_lock->events.destroy, &lock->destroy);

	server->locked = true;
	wlr_scene_node_set_enabled(&server->layer_lock->node, true);
	wlr_scene_node_raise_to_top(&server->layer_lock->node);
	update_backdrop(server);

	/* Any pointer grab a game held must not survive the lock. */
	viewport_pointer_deactivate_constraint(server);
	wlr_seat_pointer_notify_clear_focus(server->seat);

	wlr_session_lock_v1_send_locked(wlr_lock);
	wlr_log(WLR_INFO, "session locked");
}

void viewport_session_lock_init(struct viewport_server *server)
{
	server->session_lock_manager =
		wlr_session_lock_manager_v1_create(server->wl_display);
	if (server->session_lock_manager == NULL) {
		return;
	}

	server->new_session_lock.notify = handle_new_lock;
	wl_signal_add(&server->session_lock_manager->events.new_lock,
		&server->new_session_lock);
}

File: crates/viewport/src/handlers/session_lock.rs
// SPDX-License-Identifier: GPL-3.0-or-later
//
// ext-session-lock-v1. Ports src/session_lock.c.
//
// A locker is not a window and not a layer surface. It is one surface per
// output, drawn over everything, and while it is up nothing else may be seen
// or typed into — that is the protocol's whole guarantee, and it is the
// compositor's to keep rather than the shell's. The shell is a web page; a
// lock screen that a page could paint over would not be a lock screen.
//
// The other half of the guarantee is what happens when the locker dies. The
// session stays locked: the surfaces go, the screen shows nothing, and the
// only way out is another locker taking over. Unlocking on a crash would make
// killing the locker the way past it.

use smithay::reexports::wayland_server::protocol::wl_output::WlOutput;
use smithay::utils::SERIAL_COUNTER;
use smithay::wayland::session_lock::{
    LockSurface, SessionLockHandler, SessionLockManagerState, SessionLocker,
};

use crate::state::ViewportState;

impl SessionLockHandler for ViewportState {
    fn lock_state(&mut self) -> &mut SessionLockManagerState {
        &mut self.session_lock_state
    }

    fn lock(&mut self, confirmation: SessionLocker) {
        // A second locker over a working one is refused.
        //
        // Smithay grants every `lock` request — it builds a fresh lock object
        // with its own status flag and calls this, on the assumption that the
        // compositor decides — so nothing below this refuses it for us. Taking
        // it clears `lock_surfaces` a few lines down, which drops the surfaces
        // of the locker that is *still running and still drawing*: two clients
        // then own one screen, only the newer one is rendered, and unlocking
        // the one you can see leaves the other holding a lock nothing can
        // reach. That is a session that stays locked after a correct password,
        // showing the clear colour and no way back.
        if self.locked && self.lock_screen_is_drawing() {
            tracing::warn!(
                "a locker asked for a session that another locker is already \
                 drawing; refusing it. Dropping the request tells that client \
                 `finished`, which is what it is waiting to hear."
            );
            // Dropped rather than confirmed: `SessionLocker::drop` sends
            // `finished`, and swaylock exits on it.
            return;
        }

        tracing::info!("session locked");
        self.locked = true;
        self.locked_at = Some(std::time::Instant::now());
        self.lock_warned = false;
        self.lock_surfaces.clear();
        // Whatever the shell was drawing, it is not the lock screen any more.
        //
        // Reached where the built-in screen is up and not drawing — a shell
        // that hung or died — which is the case a locker is *meant* to be able
        // to take over, and the refusal above already turned away the case
        // where it is drawing. Dropping it here stops the shell's next frame
        // from being drawn over the locker that has just taken the session,
        // and tells the page to put its own lock screen away.
        if self.lock_shell_drawn.take().is_some() {
            self.notify(&viewport_ipc::Event::SessionUnlock);
        }
        self.lock_attempt = None;

        // Keyboard focus goes nowhere until a lock surface arrives and takes
        // it. Leaving it where it was would let the previously focused window
        // keep receiving keys behind the lock screen.
        if let Some(keyboard) = self.seat.get_keyboard() {
            let serial = SERIAL_COUNTER.next_serial();
            keyboard.set_focus(self, None, serial);
        }

        // Confirmed only once nothing else can be seen, which is what the
        // client waits for before it will show anything of its own.
        confirmation.lock();
        self.needs_render = true;
    }

    fn unlock(&mut self) {
        tracing::info!("session unlocked");
        self.locked = false;
        self.locked_at = None;
        self.lock_surfaces.clear();
        // The page is told either way. It will not normally have a lock screen
        // up when an external locker unlocks — the two do not run together —
        // but it does after a takeover of a built-in lock that had stopped
        // drawing, and a page left holding one over an unlocked session is a
        // desktop nobody can click.
        if self.lock_shell_drawn.take().is_some() || self.lock_mode.is_built_in() {
            self.notify(&viewport_ipc::Event::SessionUnlock);
        }
        self.lock_attempt = None;
        self.needs_render = true;
    }

    fn new_surface(&mut self, surface: LockSurface, output: WlOutput) {
        use smithay::output::Output;

        let Some(output) = Output::from_resource(&output) else {
            // Nothing to size it against, and a lock surface with no size
            // never paints — which is a locked session showing whatever was
            // on screen before, with no way back.
            tracing::error!("a lock surface arrived for an output we do not know");
            return;
        };
        let Some(geometry) = self.space.output_geometry(&output) else {
            tracing::error!(
                "a lock surface arrived for {} which is not mapped",
                output.name()
            );
            return;
        };
        tracing::info!("lock surface on {}", output.name());

        // The whole output, because that is the only size a lock surface may
        // be. It will not paint until it has been told.
        surface.with_pending_state(|state| {
            state.size = Some((geometry.size.w as u32, geometry.size.h as u32).into());
        });
        surface.send_configure();

        self.lock_surfaces.insert(output.name(), surface);
        self.needs_render = true;
    }
}

impl ViewportState {
    /// Whether a lock screen is up and on at least one screen.
    ///
    /// The question both refusals ask: a locked session whose locker has gone
    /// is one anybody may take over — that is the documented way out of a
    /// crashed locker, and `check_lock_screen` tells the user to do exactly
    /// that. A locked session that is *drawing* is not, and a second locker
    /// over it is how one ends up unreachable.
    ///
    /// It is also the fail-closed gate for the lock screen this compositor
    /// draws itself, and that is the harder half. The renderer asks this
    /// before it will put a single pixel of the shell's buffer on a locked
    /// screen, and false here is a black screen — which is an acceptable
    /// failure, where the desktop showing through is not.
    ///
    /// For the built-in screen it takes two facts, and neither alone is
    /// enough:
    ///
    /// * The shell has said, naming this lock, that it has painted the lock
    ///   screen. A message alone is not proof of a pixel: a page can send one
    ///   from a handler and then never paint. But a live process is not the
    ///   test either — a page that is running and stuck is exactly the case —
    ///   so something the page has to say is the only way to know it got as
    ///   far as building the thing.
    ///
    /// * *And* a frame has landed since it said so. The page sends `drawn`
    ///   from a double `requestAnimationFrame`, which runs strictly after the
    ///   frame the lock screen was rendered into was submitted — so any buffer
    ///   arriving after that message is that frame or a later one, and every
    ///   one of them has the lock screen in it. The frame the page committed
    ///   *before* it was told to lock is the desktop, and drawing that is
    ///   exactly the failure this guard exists to stop: a locked session
    ///   showing the bar, the window titles, and whatever the notification
    ///   centre last had in it.
    ///
    /// Both facts are dropped on anything that could invalidate them — a new
    /// lock, the shell process dying, its toplevel going away — so the answer
    /// after any of those is false until the page has drawn and said so again.
    /// A hung shell never says so, and a dead one cannot; both are black.
    ///
    /// The message is not privileged. Anything that can reach the control
    /// socket can send it, and the generation it names is broadcast rather
    /// than secret. That is deliberate and not a hole worth closing here: the
    /// socket is reachable only by processes already running as this user,
    /// which is a position from which the desktop was readable before the lock
    /// was ever taken. What this guards against is the shell being *broken*,
    /// which is the failure that actually happens.
    pub fn lock_screen_is_drawing(&self) -> bool {
        use smithay::utils::IsAlive as _;
        if self
            .lock_surfaces
            .values()
            .any(|surface| surface.wl_surface().alive())
        {
            return true;
        }
        self.lock_mode.is_built_in()
            && self.lock_shell_drawn.is_some_and(|(lock, frames)| {
                lock == self.lock_generation && self.shell_frames > frames
            })
    }

    /// Forget that the shell had drawn the lock screen.
    ///
    /// Called wherever the page that drew it stops being the page on screen:
    /// its process died, its toplevel went away, it was reloaded. The next
    /// thing rendered on a locked session is then black until the page that
    /// replaced it has drawn a lock screen of its own and said so.
    ///
    /// Separate from the lock itself on purpose. The session stays locked
    /// across every one of those — a shell that crashes must not be a way past
    /// the lock — and only what is *drawn* is withdrawn.
    pub fn forget_lock_screen(&mut self) {
        if self.lock_shell_drawn.take().is_some() {
            tracing::info!(
                "lock: the page that drew the lock screen has gone. The session \
                 stays locked and the screen goes black until something draws \
                 one again."
            );
            self.needs_render = true;
        }
    }

    /// Say so if the session is locked and nothing is drawing a lock screen.
    ///
    /// A locker that exits after taking the lock leaves the session locked
    /// with no way to authenticate — correct by the protocol, and identical to
    /// what the C build does, but from the front it is a black screen that
    /// eats every key. The only ways out are another locker or a VT switch,
    /// and neither is guessable from a screen that says nothing.
    ///
    /// Warned once per lock rather than every tick.
    pub fn check_lock_screen(&mut self) {
        if !self.locked || self.lock_warned {
            return;
        }
        let Some(at) = self.locked_at else {
            return;
        };

        // A locker that drew and then exited leaves surfaces behind that no
        // longer exist. Keeping them means rendering a dead client and never
        // noticing that nothing is on screen any more.
        use smithay::utils::IsAlive as _;
        let before = self.lock_surfaces.len();
        self.lock_surfaces
            .retain(|_, surface| surface.wl_surface().alive());
        if self.lock_surfaces.len() != before {
            self.needs_render = true;
        }

        if at.elapsed() < std::time::Duration::from_secs(3) {
            return;
        }

        // The built-in screen has no per-output surfaces to count — it is one
        // buffer across the whole layout — so what stands in for "nothing has
        // drawn" is the same gate the renderer asks before it draws any of it.
        //
        // The advice differs too, and that is the point of saying it
        // separately: an external locker that never draws is a program that
        // crashed, and the answer is to run another one. A shell that never
        // draws is the desktop itself, and the answer is `idle.lock_command` —
        // a locker that is not this shell, for a machine whose shell will not
        // paint.
        if self.lock_mode.is_built_in() {
            if self.lock_screen_is_drawing() {
                return;
            }
            self.lock_warned = true;
            tracing::error!(
                "locked, but the shell has not drawn a lock screen. The session \
                 stays locked — a shell that crashes must not be a way past the \
                 lock — so the way out is Ctrl+Alt+F1..F12 to another VT. Set \
                 idle.lock_command if this machine should lock with a locker of \
                 its own instead."
            );
            return;
        }

        let missing: Vec<String> = self
            .space
            .outputs()
            .map(|output| output.name())
            .filter(|name| !self.lock_surfaces.contains_key(name))
            .collect();
        if missing.is_empty() {
            return;
        }
        self.lock_warned = true;
        tracing::error!(
            "locked, but nothing has drawn a lock screen on {}. \
             The locker has probably exited. The session stays locked — that is \
             what the protocol asks for — so the way out is Ctrl+Alt+F1..F12 to \
             another VT, or running another locker against this display.",
            missing.join(", ")
        );
    }

    /// Focus a lock surface, so the locker can be typed into.
    ///
    /// Called when one commits: focusing at `new_surface` would be too early,
    /// since the client has not acknowledged its size and has nothing to show.
    pub fn focus_lock_surface(
        &mut self,
        surface: &smithay::reexports::wayland_server::protocol::wl_surface::WlSurface,
    ) {
        if !self.locked {
            return;
        }
        let ours = self
            .lock_surfaces
            .values()
            .any(|lock| lock.wl_surface() == surface);
        if !ours {
            return;
        }
        let already = self
            .seat
            .get_keyboard()
            .and_then(|keyboard| keyboard.current_focus())
            .map(|focused| focused.is_surface(surface))
            .unwrap_or(false);
        if already {
            return;
        }
        if let Some(keyboard) = self.seat.get_keyboard() {
            let serial = SERIAL_COUNTER.next_serial();
            keyboard.set_focus(self, Some(surface.clone().into()), serial);
        }
        self.needs_render = true;
    }
}
